File: tests/unit/python/test_ephemeral_defer.py
"""Manifest-driven ephemeral defer (T3.2) — Python layer.

Covers the object route form in ``scripts/lib/discord_modules.py`` (parsing,
validation, the ``ephemeral_defer_routes`` merged-list helper) and the
``DISCORD_EPHEMERAL_DEFER_ROUTES`` emission in
``scripts/generate-terraform-modules.py``.

All module fixtures live in pytest tmp dirs (``make_module`` from conftest).
The generator is exercised through its ``generate(repo_root)`` function, which
already takes the repository root as a parameter — no repo-scanning override
mechanism is needed for testability.
"""

import importlib.util
from pathlib import Path

import pytest

from discord_modules import (
    ModuleError,
    ephemeral_defer_routes,
    route_map,
    validate_module_manifests,
)

REPO_ROOT = Path(__file__).resolve().parents[3]
GENERATOR = REPO_ROOT / "scripts" / "generate-terraform-modules.py"


def load_generator():
    spec = importlib.util.spec_from_file_location("generate_terraform_modules", GENERATOR)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


# ---------------------------------------------------------------------------
# route_map: object route form
# ---------------------------------------------------------------------------


def test_route_map_normalizes_object_form_to_lambda_name(repo_root, make_module):
    make_module(
        "demo",
        {
            "name": "demo",
            "routes": {
                "commands": {
                    "account link": {
                        "lambda": "discord-cmd-account-link",
                        "ephemeral_defer": True,
                    }
                }
            },
        },
    )
    assert route_map(repo_root, "commands") == {
        "account link": "discord-cmd-account-link"
    }


def test_route_map_mixes_string_and_object_forms(repo_root, make_module):
    make_module(
        "demo",
        {
            "name": "demo",
            "routes": {
                "commands": {
                    "account link": {
                        "lambda": "discord-cmd-account-link",
                        "ephemeral_defer": True,
                    },
                    "account list": "discord-cmd-account-list",
                }
            },
        },
    )
    assert route_map(repo_root, "commands") == {
        "account link": "discord-cmd-account-link",
        "account list": "discord-cmd-account-list",
    }


# ---------------------------------------------------------------------------
# ephemeral_defer_routes: merged, sorted, deduplicated
# ---------------------------------------------------------------------------


def test_ephemeral_defer_routes_merges_sorted_and_deduped():
    manifests = [
        {
            "name": "m1",
            "routes": {
                "commands": {
                    "example ping": {"lambda": "discord-cmd-example-ping", "ephemeral_defer": True},
                    "zeta cmd": {"lambda": "discord-cmd-zeta-cmd", "ephemeral_defer": True},
                }
            },
        },
        {
            "name": "m2",
            "routes": {
                "commands": {
                    # Duplicate of m1's entry — must be deduplicated.
                    "example ping": {"lambda": "discord-cmd-example-ping", "ephemeral_defer": True},
                    "account link": {"lambda": "discord-cmd-account-link", "ephemeral_defer": True},
                }
            },
        },
    ]
    assert ephemeral_defer_routes(manifests) == [
        "account link",
        "example ping",
        "zeta cmd",
    ]


def test_ephemeral_defer_routes_string_and_false_forms_contribute_nothing():
    manifests = [
        {
            "name": "m1",
            "routes": {
                "commands": {
                    "plain string": "discord-cmd-plain-string",
                    "explicit false": {
                        "lambda": "discord-cmd-explicit-false",
                        "ephemeral_defer": False,
                    },
                    "no flag": {"lambda": "discord-cmd-no-flag"},
                },
                # Only command routes may opt in; other kinds never contribute.
                "components": {"pager": "discord-component-pager"},
            },
        }
    ]
    assert ephemeral_defer_routes(manifests) == []


def test_ephemeral_defer_routes_empty_manifest_list():
    assert ephemeral_defer_routes([]) == []


def test_ephemeral_defer_routes_collects_context_menu_kinds():
    """Finding 2: user_commands / message_commands may opt in, keyed by the raw
    command name (the ingress matches context menus on the raw name)."""
    manifests = [
        {
            "name": "m1",
            "routes": {
                "commands": {
                    "account link": {
                        "lambda": "discord-cmd-account-link",
                        "ephemeral_defer": True,
                    }
                },
                "user_commands": {
                    "Report User": {
                        "lambda": "discord-usercmd-report-user",
                        "ephemeral_defer": True,
                    },
                    "Block User": {"lambda": "discord-usercmd-block-user"},
                },
                "message_commands": {
                    "Pin Message": {
                        "lambda": "discord-msgcmd-pin-message",
                        "ephemeral_defer": True,
                    }
                },
            },
        }
    ]
    assert ephemeral_defer_routes(manifests) == [
        "Pin Message",
        "Report User",
        "account link",
    ]


# ---------------------------------------------------------------------------
# Validation of the object route form
# ---------------------------------------------------------------------------


def test_validation_accepts_object_form_command_route(repo_root, make_module):
    make_module(
        "demo",
        {
            "name": "demo",
            "routes": {
                "commands": {
                    "account link": {
                        "lambda": "discord-cmd-account-link",
                        "ephemeral_defer": True,
                    }
                }
            },
        },
    )
    validate_module_manifests(repo_root)  # must not raise


def test_validation_rejects_ephemeral_defer_on_component_routes(repo_root, make_module):
    make_module(
        "demo",
        {
            "name": "demo",
            "routes": {
                "components": {
                    "pager": {
                        "lambda": "discord-component-pager",
                        "ephemeral_defer": True,
                    }
                }
            },
        },
    )
    with pytest.raises(ModuleError, match="ephemeral_defer"):
        validate_module_manifests(repo_root)


def test_validation_accepts_ephemeral_defer_on_user_commands(repo_root, make_module):
    # Finding 2: context-menu route kinds may carry the flag.
    make_module(
        "demo",
        {
            "name": "demo",
            "routes": {
                "user_commands": {
                    "Report User": {
                        "lambda": "discord-usercmd-report-user",
                        "ephemeral_defer": True,
                    }
                }
            },
        },
    )
    validate_module_manifests(repo_root)  # must not raise


def test_validation_accepts_ephemeral_defer_on_message_commands(repo_root, make_module):
    make_module(
        "demo",
        {
            "name": "demo",
            "routes": {
                "message_commands": {
                    "Pin Message": {
                        "lambda": "discord-msgcmd-pin-message",
                        "ephemeral_defer": True,
                    }
                }
            },
        },
    )
    validate_module_manifests(repo_root)  # must not raise


def test_validation_rejects_non_boolean_ephemeral_defer(repo_root, make_module):
    make_module(
        "demo",
        {
            "name": "demo",
            "routes": {
                "commands": {
                    "account link": {
                        "lambda": "discord-cmd-account-link",
                        "ephemeral_defer": "yes",
                    }
                }
            },
        },
    )
    with pytest.raises(ModuleError, match="ephemeral_defer.*boolean"):
        validate_module_manifests(repo_root)


def test_validation_rejects_object_route_missing_lambda(repo_root, make_module):
    make_module(
        "demo",
        {
            "name": "demo",
            "routes": {"commands": {"account link": {"ephemeral_defer": True}}},
        },
    )
    with pytest.raises(ModuleError, match="lambda"):
        validate_module_manifests(repo_root)


def test_validation_still_rejects_non_string_non_object_route_values(repo_root, make_module):
    make_module(
        "demo",
        {"name": "demo", "routes": {"commands": {"account link": 42}}},
    )
    with pytest.raises(ModuleError):
        validate_module_manifests(repo_root)


# ---------------------------------------------------------------------------
# Generator: DISCORD_EPHEMERAL_DEFER_ROUTES emission in generated Terraform
# ---------------------------------------------------------------------------


def test_generator_emits_merged_csv_local(repo_root, make_module):
    make_module(
        "m1",
        {
            "name": "m1",
            "routes": {
                "commands": {
                    "example ping": {
                        "lambda": "discord-cmd-example-ping",
                        "ephemeral_defer": True,
                    }
                }
            },
        },
    )
    make_module(
        "m2",
        {
            "name": "m2",
            "routes": {
                "commands": {
                    "account link": {
                        "lambda": "discord-cmd-account-link",
                        "ephemeral_defer": True,
                    },
                    "account list": "discord-cmd-account-list",
                }
            },
        },
    )
    generator = load_generator()
    files = generator.generate(repo_root)
    modules_tf = files[repo_root / "infra" / "terraform" / "generated_modules.tf"]
    assert 'discord_ephemeral_defer_routes = "account link,example ping"' in modules_tf


def test_generator_emits_context_menu_raw_names_in_csv(repo_root, make_module):
    # Finding 2: opted-in context-menu commands (matched by raw name) merge into
    # the generated DISCORD_EPHEMERAL_DEFER_ROUTES list alongside command paths.
    make_module(
        "m1",
        {
            "name": "m1",
            "routes": {
                "commands": {
                    "account link": {
                        "lambda": "discord-cmd-account-link",
                        "ephemeral_defer": True,
                    }
                },
                "user_commands": {
                    "Report User": {
                        "lambda": "discord-usercmd-report-user",
                        "ephemeral_defer": True,
                    }
                },
            },
        },
    )
    generator = load_generator()
    files = generator.generate(repo_root)
    modules_tf = files[repo_root / "infra" / "terraform" / "generated_modules.tf"]
    assert 'discord_ephemeral_defer_routes = "Report User,account link"' in modules_tf


def test_generator_empty_case_emits_no_stale_value(repo_root, make_module):
    make_module(
        "m1",
        {
            "name": "m1",
            "routes": {"commands": {"example ping": "discord-cmd-example-ping"}},
        },
    )
    generator = load_generator()
    files = generator.generate(repo_root)
    modules_tf = files[repo_root / "infra" / "terraform" / "generated_modules.tf"]
    # The local must exist (the root main.tf references it unconditionally) but
    # carry an empty value — never a stale route list.
    assert 'discord_ephemeral_defer_routes = ""' in modules_tf
    assert "example ping" not in modules_tf.split("discord_ephemeral_defer_routes")[1].split("\n")[0]


def test_root_main_tf_wires_generated_local_into_ingress_env():
    main_tf = (REPO_ROOT / "infra" / "terraform" / "main.tf").read_text(encoding="utf-8")
    assert "DISCORD_EPHEMERAL_DEFER_ROUTES" in main_tf
    assert "local.discord_ephemeral_defer_routes" in main_tf

File: tests/unit/cpp/test_ephemeral_defer.cpp
// T3.2 — manifest-driven ephemeral defer at ingress.
//
// Covers the two pure C++ pieces of the chain:
//   - discord_interactions::route_in_csv_allowlist: the ingress-side parser
//     for the generated DISCORD_EPHEMERAL_DEFER_ROUTES env var (comma-separated
//     command paths, entries whitespace-trimmed, exact full-path match).
//   - discord_interactions::route_from_manifest_entry: accepts both the plain
//     string route form ("discord-cmd-x") and the object route form
//     ({"lambda": "discord-cmd-x", "ephemeral_defer": true}).

#include <doctest/doctest.h>

#include <nlohmann/json.hpp>

#include <stdexcept>

#include "discord_interactions/interaction.hpp"
#include "discord_interactions/routing.hpp"

using nlohmann::json;
using discord_interactions::route_in_csv_allowlist;
using discord_interactions::route_from_manifest_entry;

// ---------------------------------------------------------------------------
// route_in_csv_allowlist
// ---------------------------------------------------------------------------

TEST_CASE("route_in_csv_allowlist: empty csv (unset env) matches nothing") {
    CHECK_FALSE(route_in_csv_allowlist("account link", ""));
}

TEST_CASE("route_in_csv_allowlist: single entry exact match") {
    CHECK(route_in_csv_allowlist("account link", "account link"));
}

TEST_CASE("route_in_csv_allowlist: multiple entries, match anywhere") {
    const std::string csv{"example ping,account link,other cmd"};
    CHECK(route_in_csv_allowlist("example ping", csv));
    CHECK(route_in_csv_allowlist("account link", csv));
    CHECK(route_in_csv_allowlist("other cmd", csv));
}

TEST_CASE("route_in_csv_allowlist: entries are whitespace-trimmed") {
    CHECK(route_in_csv_allowlist("account link", "  account link  , other cmd "));
    CHECK(route_in_csv_allowlist("other cmd", "  account link  , other cmd "));
    CHECK(route_in_csv_allowlist("account link", "\taccount link\t"));
}

TEST_CASE("route_in_csv_allowlist: no match") {
    CHECK_FALSE(route_in_csv_allowlist("account unlink", "account link,other cmd"));
}

TEST_CASE("route_in_csv_allowlist: exact full-path match only") {
    // "account" must NOT match the "account link" entry, and vice versa.
    CHECK_FALSE(route_in_csv_allowlist("account", "account link"));
    CHECK_FALSE(route_in_csv_allowlist("account link", "account"));
    CHECK_FALSE(route_in_csv_allowlist("account link extra", "account link"));
    // Interior whitespace is part of the path; only edges are trimmed.
    CHECK_FALSE(route_in_csv_allowlist("account  link", "account link"));
}

TEST_CASE("route_in_csv_allowlist: empty command path never matches") {
    CHECK_FALSE(route_in_csv_allowlist("", "account link"));
    CHECK_FALSE(route_in_csv_allowlist("", ","));
    CHECK_FALSE(route_in_csv_allowlist("", " , "));
}

TEST_CASE("route_in_csv_allowlist: empty and whitespace-only entries are skipped") {
    CHECK(route_in_csv_allowlist("account link", ",account link,"));
    CHECK(route_in_csv_allowlist("account link", " , account link , "));
    CHECK_FALSE(route_in_csv_allowlist("account link", ", ,"));
}

// ---------------------------------------------------------------------------
// route_from_manifest_entry: string + object route forms
// ---------------------------------------------------------------------------

namespace {

json manifest_with_command_route(const json& route_value) {
    return json{
        {"name", "example"},
        {"routes", {{"commands", {{"account link", route_value}}}}},
        {"error_mapper", {{"lambda", "discord-error-mapper-example"}}},
    };
}

}  // namespace

TEST_CASE("route_from_manifest_entry: plain string form still works") {
    const json manifest = manifest_with_command_route("discord-cmd-account-link");
    const auto route = route_from_manifest_entry(manifest, "commands", "account link");
    CHECK(route.module == "example");
    CHECK(route.route == "account link");
    CHECK(route.function_name == "discord-cmd-account-link");
    CHECK(route.error_mapper_function_name == "discord-error-mapper-example");
}

TEST_CASE("route_from_manifest_entry: object form with ephemeral_defer") {
    const json manifest = manifest_with_command_route(
        json{{"lambda", "discord-cmd-account-link"}, {"ephemeral_defer", true}});
    const auto route = route_from_manifest_entry(manifest, "commands", "account link");
    CHECK(route.function_name == "discord-cmd-account-link");
    CHECK(route.route == "account link");
}

TEST_CASE("route_from_manifest_entry: object form without ephemeral_defer") {
    const json manifest = manifest_with_command_route(
        json{{"lambda", "discord-cmd-account-link"}});
    const auto route = route_from_manifest_entry(manifest, "commands", "account link");
    CHECK(route.function_name == "discord-cmd-account-link");
}

TEST_CASE("route_from_manifest_entry: object form missing lambda key throws") {
    const json manifest = manifest_with_command_route(json{{"ephemeral_defer", true}});
    CHECK_THROWS_AS(
        route_from_manifest_entry(manifest, "commands", "account link"),
        std::runtime_error);
}

TEST_CASE("route_from_manifest_entry: object form with non-string lambda throws") {
    const json manifest = manifest_with_command_route(json{{"lambda", 42}});
    CHECK_THROWS_AS(
        route_from_manifest_entry(manifest, "commands", "account link"),
        std::runtime_error);
}

TEST_CASE("route_from_manifest_entry: missing route still throws") {
    const json manifest = manifest_with_command_route("discord-cmd-account-link");
    CHECK_THROWS_AS(
        route_from_manifest_entry(manifest, "commands", "no such route"),
        std::runtime_error);
}
